### PR TITLE
feat: add public ip address support for `/res/api-management/service`

### LINKED ISSUE #1104: [AVM Module Issue]: Public ip adress for API Management
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/api-management/service

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

At this moment the property 'publicIpAddressId' is not exposed in this module.

Creating a APIM with the virtualNetworkType set to 'Internal' without specifying a public ip adress will result in a APIM with a platformVersion stv1 to be created. stv1 will retire on august 2024.
[stv1-platform-retirement-august-2024](https://learn.microsoft.com/en-us/azure/api-management/breaking-changes/stv1-platform-retirement-august-2024)

Can you expose the property 'publicIpAddressId' to work around this issue.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by tony-box**
## Description

Closes #1104 

Adds APIM support for PublicIPAddress

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.api-management.service](https://github.com/tony-box/bicep-registry-modules/actions/workflows/avm.res.api-management.service.yml)      |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: .github/workflows/avm.ptn.security-center.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+name: "avm.ptn.security-center"
```
Comment (AlexanderSehr) [inline .github/workflows/avm.ptn.security-center.yml:1]: Beware, PR invaders

File: avm/res/api-management/service/README.md [hunk 2]
```diff
@@ -1532,39 +1551,6 @@ A list of availability zones denoting where the resource needs to come from.
 
 _None_
 
-## Notes
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/README.md:1535]: Just doublecheckinf if this was intended. You're more than welcome to remove these notes that originally came from CARML. Just want to make sure you did know about them 😉

File: avm/res/api-management/service/main.bicep [hunk 3]
```diff
@@ -141,7 +143,7 @@ var identity = !empty(managedIdentities)
   ? {
       type: (managedIdentities.?systemAssigned ?? false)
         ? (!empty(managedIdentities.?userAssignedResourceIds ?? {}) ? 'SystemAssigned,UserAssigned' : 'SystemAssigned')
-        : (!empty(managedIdentities.?userAssignedResourceIds ?? {}) ? 'UserAssigned' : 'None')
+        : (!empty(managedIdentities.?userAssignedResourceIds ?? {}) ? 'UserAssigned' : null)
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/main.bicep:146]: Why this change? It seems to bring back the RP warning 
`/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/api-management/service/main.bicep(142,16) : Warning BCP321: Expected a value of type "string" but the provided value is of type "'SystemAssigned' | 'SystemAssigned,UserAssigned' | 'UserAssigned' | null".`

File: avm/res/api-management/service/tests/e2e/defaults/main.test.bicep [hunk 4]
```diff
@@ -36,15 +36,13 @@ resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
...
-  for iteration in ['init', 'idem']: {
-    scope: resourceGroup
-    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
-    params: {
-      name: '${namePrefix}${serviceShort}001'
-      location: resourceLocation
-      publisherEmail: 'apimgmt-noreply@mail.windowsazure.com'
-      publisherName: '${namePrefix}-az-amorg-x-001'
-    }
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/defaults/main.test.bicep:39]: Beware, you're not using the latest Bicep version which is why every Bicep file (& JSON file) got re-formatted / updated.

Please upgrade to the latest (check `bicep --version`) and run `Set-AVMModule -Recurse` on the root folder path of this module again :)

File: avm/res/api-management/service/tests/e2e/max/dependencies.bicep [hunk 5]
```diff
@@ -4,13 +4,219 @@ param managedIdentityName string
...
+param vnetName string = 'VNet'
+
+@description('Address prefix')
+param vnetAddressPrefix string = '10.0.0.0/16'
+
+@description('Subnet Prefix')
+param subnetPrefix string = '10.0.0.0/24'
+
+@description('Subnet Name')
+param subnetName string = 'Subnet'
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/max/dependencies.bicep:20]: This and the next is just a recommendation to align with other test cases. Feel free to reject if you prefer the original implementation 😉 

```suggestion
@description('Required. The name of the Virtual Network to create.')
param virtualNetworkName string

var addressPrefix = '10.0.0.0/16'
```

File: avm/res/api-management/service/tests/e2e/max/dependencies.bicep [hunk 6]
```diff
@@ -4,13 +4,219 @@ param managedIdentityName string
...
+
+@description('DNS Prefix')
+param dnsLabelPrefix string
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/max/dependencies.bicep:23]: This and the next is just a recommendation to align with other test cases. Feel free to reject if you prefer the original implementation 😉 

```suggestion
@description('Required. The name of the Virtual Network to create.')
param virtualNetworkName string

var addressPrefix = '10.0.0.0/16'
```

File: avm/res/api-management/service/tests/e2e/max/dependencies.bicep [hunk 7]
```diff
@@ -4,13 +4,219 @@ param managedIdentityName string
...
+            }
+          ]
+        }
+      }
+    ]
+  }
+}
+
+resource nsg 'Microsoft.Network/networkSecurityGroups@2020-06-01' = {
+  name: 'testNSG'
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/max/dependencies.bicep:65]: For rationale, see other comment.

```suggestion
resource vnet 'Microsoft.Network/virtualNetworks@2023-04-01' = {
  name: vnetName
  location: location
  properties: {
    addressSpace: {
      addressPrefixes: [
        addressPrefix
      ]
    }
    subnets: [
      {
        name: 'default'
        properties: {
          addressPrefix: cidrSubnet(addressPrefix, 24, 0)
          networkSecurityGroup: {
            id: nsg.id
          }
          serviceEndpoints: [
            {
              service: 'Microsoft.Storage'
            }
            {
              service: 'Microsoft.Sql'
            }
            {
              service: 'Microsoft.KeyVault'
            }
          ]
        }
      }
    ]
  }
}

resource nsg 'Microsoft.Network/networkSecurityGroups@2020-06-01' = {
  name: networkSecurityGroupName
```

File: avm/res/api-management/service/tests/e2e/max/dependencies.bicep [hunk 8]
```diff
@@ -4,13 +4,219 @@ param managedIdentityName string
...
+  location: location
+  sku: {
+    name: 'Standard'
+    tier: 'Regional'
+  }
+  properties: {
+    publicIPAllocationMethod: 'Static'
+    publicIPAddressVersion: 'IPv4'
+    dnsSettings: {
+      domainNameLabel: dnsLabelPrefix
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/max/dependencies.bicep:207]: See other comment

```suggestion
      domainNameLabel: publicIpDnsLabelPrefix
```

File: avm/res/api-management/service/tests/e2e/max/main.test.bicep [hunk 9]
```diff
@@ -41,6 +41,8 @@ module nestedDependencies 'dependencies.bicep' = {
   params: {
     managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
     location: resourceLocation
+    publicIPName: 'dep-${namePrefix}-pip-${serviceShort}'
+    dnsLabelPrefix: 'dep-${namePrefix}-dnsprefix-${uniqueString(deployment().name, resourceLocation)}'
```
Comment (AlexanderSehr) [inline avm/res/api-management/service/tests/e2e/max/main.test.bicep:45]:
```suggestion
    publicIpDnsLabelPrefix: 'dep-${namePrefix}-dnsprefix-${uniqueString(deployment().name, resourceLocation)}'
```

**Comment by tony-box** [comment]
Been a bit busy, but thank you for all the recommendations @AlexanderSehr -- I didn't do a thorough review of the way this was set up during the CARML -> AVM migraiton so there are still a lot of relics from CARML. I will make the changes you suggested. I'm still working on fixing the max test deployment which includes creating a publicIP. have been running into issues with it and have even raised an Azure support ticket. I'm still working on my branch so there will be more updates coming

**Comment by tony-box** [comment]
I am going to re-do this whole pull request now that I have my local branches sorted out and the commits cleaned up. Stay tuned...

**Review by eriqua — COMMENTED** [review]
Hey @tony-box is this PR still valid or should we close it as superseded?

**Comment by tony-box** [comment]
closing this PR which will be superceeded by #2453
